Document new getStaticPaths type helpers (#2765)
withastro/docs: src/content/docs/en/guides/typescript.mdx

```
:::

> ### getStaticPaths

Reviewer: Not quite happy with this but two aims here:
1) Small style nit for `getStaticPaths()` being in backticks and including parens.
2) I’m trying to communicate that this is not the main `getStaticPaths()` docs in the heading — important for search results — while not making this too verbose in context on the page

```suggestion
### Infer `getStaticPaths()` types
```

<Since v="2.1.0" />

Astro includes helpers for working with the types returned in your [getStaticPaths()](/en/reference/api-reference/#getstaticpaths) export for dynamic pages.

> To get the type of [params](/en/reference/api-reference/#params) you can do the following:
> 
> ```astro title="src/pages/posts[...slug].astro" "@components" "@layouts"
> ---
> import { InferGetStaticParamsType } from 'astro';
> 
> export async function getStaticPaths() {
>   const results = await fetchPosts();
>   return results.map(entry => {
>     return {
>       params: { slug: entry.slug },
>     }
>   });
> }
> 
> type Params = InferGetStaticParamsType<typeof getStaticPaths>;
> //   ^? { slug: string; }
> ---
> ```
> 
> Likewise, to get the [props](/en/reference/api-reference/#astroprops) for the page, use the `InferGetStaticPropsType` type function:
> 
> ```astro title="src/pages/posts[...slug].astro" "@components" "@layouts"
> ---
> import { InferGetStaticPropsType } from 'astro';
> 
> export async function getStaticPaths() {
>   const results = await fetchPosts();
>   return results.map(entry => {
>     return {
>       params: { slug: entry.slug },
>       props: {
>         propA: true,
>         propB: 22
>       }
>     }
>   });
> }
> 
> type Props = InferGetStaticPropsType<typeof getStaticPaths>;
> // ^? { propA: boolean; propB: number; }
> ---
> ```

Reviewer: Do you think it would make sense to combine these two examples? They share a lot and you’ll often use both together.

I’ve also tweaked the example to use the real-world `getCollection()` call instead of a hypothetical `fetchPosts()` and moved to show the types on `Astro.props`/`params` to help make it clear that setting types here allows Astro to infer the type of those.

````suggestion
You can get the type of [`Astro.params`](/en/reference/api-reference/#astroparams) with `InferGetStaticParamsType` and the type of [`Astro.props`](/en/reference/api-reference/#astroprops) with `InferGetStaticPropsType`:

```astro title="src/pages/posts/[...slug].astro" {2,14-15}
---
import { InferGetStaticParamsType, InferGetStaticPropsType } from 'astro';

export async function getStaticPaths() {
  const posts = await getCollection('blog');
  return posts.map((post) => {
    return {
      params: { slug: post.slug },
      props: { draft: post.data.draft, title: post.data.title },
    };
  });
}

type Params = InferGetStaticParamsType<typeof getStaticPaths>;
type Props = InferGetStaticPropsType<typeof getStaticPaths>;

const { slug } = Astro.params;
//               ^? { slug: string; }
const { title } = Astro.props;
//                ^? { draft: boolean; title: string; }
---
```
````

Author: yep! like this.
